src/topology/continuous_function/zero_at_infty.lean

```diff
@@ -0,0 +1,318 @@
+
+lemma eq_of_empty [is_empty α] (f g : α →C₀ β) : f = g :=
+ext $ is_empty.elim ‹_›
+
+/-- A continuous function on a compact space is automatically a continuous function vanishing at
+infitnity. -/
+@[simps]
+def continuous_map.lift_zero_at_infty [compact_space α] : C(α, β) ≃ (α →C₀ β) :=
+{ to_fun := λ f, { to_fun := f, continuous_to_fun := f.continuous, zero_at_infty' := by simp },
+  inv_fun := λ f, f,
+  left_inv := λ f, by { ext, refl },
+  right_inv := λ f, by { ext, refl } }
+
+/-- A continuous function on a compact space is automatically a continuous function vanishing at
+infitnity. This is not an instance to avoid type class loops. -/
+def zero_at_infty_continuous_map_class.of_compact {G : Type*} [continuous_map_class G α β]
+  [compact_space α] : zero_at_infty_continuous_map_class G α β :=
+{ coe := λ g, g,
+  coe_injective' := λ f g h, fun_like.coe_fn_eq.mp h,
+  map_continuous := map_continuous,
+  zero_at_infty := by simp }
+
+end basics
+
+section algebraic_structure
+
+variables [topological_space β] (x : α)
+
+instance [has_zero β] : has_zero (α →C₀ β) := ⟨⟨0, tendsto_const_nhds⟩⟩
+
+instance [has_zero β] : inhabited (α →C₀ β) := ⟨0⟩
+
+@[simp] lemma coe_zero [has_zero β] : ⇑(0 : α →C₀ β) = 0 := rfl
+lemma zero_apply [has_zero β] : (0 : α →C₀ β) x = 0 := rfl
+
+instance [mul_zero_class β] [has_continuous_mul β] : has_mul (α →C₀ β) :=
+⟨λ f g, ⟨f * g, by simpa only [mul_zero] using ((zero_at_infty f).mul (zero_at_infty g) :
+  tendsto (λ x : α, f x * g x) (cocompact α) (𝓝 (0 * 0)))⟩⟩
+
+@[simp] lemma coe_mul [mul_zero_class β] [has_continuous_mul β] (f g : α →C₀ β) :
+  ⇑(f * g) = f * g := rfl
+lemma mul_apply [mul_zero_class β] [has_continuous_mul β] (f g : α →C₀ β) :
+  (f * g) x = f x * g x := rfl
+
+instance [mul_zero_class β] [has_continuous_mul β] : mul_zero_class (α →C₀ β) :=
+fun_like.coe_injective.mul_zero_class _ coe_zero coe_mul
+
+instance [semigroup_with_zero β] [has_continuous_mul β] : semigroup_with_zero (α →C₀ β) :=
+fun_like.coe_injective.semigroup_with_zero _ coe_zero coe_mul
+
+instance [add_zero_class β] [has_continuous_add β] : has_add (α →C₀ β) :=
+⟨λ f g, ⟨f + g, by simpa only [add_zero] using ((zero_at_infty f).add (zero_at_infty g) :
+  tendsto (λ x : α, f x + g x) (cocompact α) (𝓝 (0 + 0)))⟩⟩
+
+@[simp] lemma coe_add [add_zero_class β] [has_continuous_add β] (f g : α →C₀ β) :
+  ⇑(f + g) = f + g := rfl
+lemma add_apply [add_zero_class β] [has_continuous_add β] (f g : α →C₀ β) :
+  (f + g) x = f x + g x := rfl
+
+instance [add_zero_class β] [has_continuous_add β] : add_zero_class (α →C₀ β) :=
+fun_like.coe_injective.add_zero_class _ coe_zero coe_add
+
+section add_monoid
+
+variables [add_monoid β] [has_continuous_add β] (f g : α →C₀ β)
+
+@[simp] lemma coe_nsmul_rec : ∀ n, ⇑(nsmul_rec n f) = n • f
+| 0 := by rw [nsmul_rec, zero_smul, coe_zero]
+| (n + 1) := by rw [nsmul_rec, succ_nsmul, coe_add, coe_nsmul_rec]
+
+instance has_nat_scalar : has_scalar ℕ (α →C₀ β) :=
```

Comment:
I believe you can skip this instance and the corresponding `int` one, and it will be made for free by the `add_monoid` (resp `add_group`) instances.  Any evidence pro/con?

Reply:
You need them if you use the `fun_like.coe_injective.add_monoid` approach like I did (which I believe is easiest, despite the explicit `has_scalar ℕ` instance). Note that the last argument to the `add_monoid` instance, namely `(λ _ _, rfl)`, is actually `coe_nsmul` (but I can't add that lemma because `simp_nf` linter complains).